src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,163 @@
+            // encoding. This would allow the char -> byte conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            _maxByteBufferSize = Math.Min(MaxByteBufferSize, targetEncoding.GetMaxByteCount(MaxCharBufferSize));
+
+            _decoder = Encoding.UTF8.GetDecoder();
+            _encoder = targetEncoding.GetEncoder();
+        }
+
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+            => _stream.FlushAsync(cancellationToken);
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+            => throw new NotSupportedException();
+
+        public override void SetLength(long value)
+            => throw new NotSupportedException();
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentException(SR.Argument_InvalidOffLen);
+            }
+
+            var bufferSegment = new ArraySegment<byte>(buffer, offset, count);
+            return WriteAsyncCore(bufferSegment, cancellationToken);
+        }
+
+        private async Task WriteAsyncCore(ArraySegment<byte> bufferSegment, CancellationToken cancellationToken)
+        {
+            bool decoderCompleted = false;
+
+            while (!decoderCompleted)
+            {
+                _decoder.Convert(bufferSegment.Array!, bufferSegment.Offset, bufferSegment.Count, _charBuffer, _charsDecoded, _charBuffer.Length - _charsDecoded,
+                    flush: false, out int bytesDecoded, out int charsDecoded, out decoderCompleted);
+
+                _charsDecoded += charsDecoded;
+                bufferSegment = bufferSegment.Slice(bytesDecoded);
+
+                if (!decoderCompleted)
```

Comment:
No need to put this inside an _if_ block. The decoder will always make as much progress as possible, so it's optimal to call `WriteBufferAsync` unconditionally.

Reply:
I don't quite understand what the original author was trying to achieve by skipping `WriteBufferAsync` on the last decoding but I assume that if we remove this condition, we no longer need to call `WriteBufferAsync` on `FinalWriteAsync`.

cc @pranavkm